### ISSUE TITLE
createFrameAtIndex has a bug 

### ISSUE THREAD

**Issue by AKZHW** [SDWebImage/SDWebImage#3351]
"  decodingOptions[(__bridge NSString *)kCGImageSourceCreateThumbnailWithTransform] = @(preserveAspectRatio);
        CGFloat maxPixelSize;
        if (preserveAspectRatio) {
            CGFloat pixelRatio = pixelWidth / pixelHeight ;
            CGFloat thumbnailRatio = thumbnailSize.width / thumbnailSize.height;
            if (pixelRatio > thumbnailRatio) {
                maxPixelSize = thumbnailSize.width;
            } else {
                maxPixelSize = thumbnailSize.height;
            }
        } else {
            maxPixelSize = MAX(thumbnailSize.width, thumbnailSize.height);
        }
        decodingOptions[(__bridge NSString *)kCGImageSourceThumbnailMaxPixelSize] = @(maxPixelSize);
        decodingOptions[(__bridge NSString *)kCGImageSourceCreateThumbnailFromImageAlways] = @(YES);
        imageRef = CGImageSourceCreateThumbnailAtIndex(source, index, (__bridge CFDictionaryRef)[decodingOptions copy]);"
        
        
        
There is a precision loss problem in this line ” CGFloat pixelRatio = pixelWidth / pixelHeight ; “

then "
 
            if (pixelRatio > thumbnailRatio) {
                maxPixelSize = thumbnailSize.width;
            } else {
                maxPixelSize = thumbnailSize.height;
            }"
should after Calculate the width and height in the same proportion, and then calculate maxsize

**Comment by dreampiggy** [comment]
Maybe https://github.com/SDWebImage/SDWebImage/pull/3262 already solve this.

**Comment by AKZHW** [comment]
only fix  precision loss problem, but calculate maxPixelSize has a problem,

<img width="913" alt="image" src="https://user-images.githubusercontent.com/9245326/172596042-714eec79-ef5c-4770-830a-5a0b75dc2f78.png">

I  think should use this code 
        CGFloat maxPixelSize;
        if (preserveAspectRatio) {
            CGFloat pixelRatio = pixelWidth / pixelHeight;
            CGFloat thumbnailRatio = thumbnailSize.width / thumbnailSize.height;
            if (pixelRatio > thumbnailRatio) {
                maxPixelSize = MAX(thumbnailSize.width, thumbnailSize.width / pixelRatio);
            } else {
                maxPixelSize = MAX(thumbnailSize.height * pixelRatio, thumbnailSize.height);
            }
        } else {
            maxPixelSize = MAX(thumbnailSize.width, thumbnailSize.height);
        }

**Comment by AKZHW** [comment]
for example  my picture origin size 360 * 136 , request size 200 * 80， the result should get 200 * 75， bug use you code get 80 * 30

**Comment by dreampiggy** [comment]
I think you're right.

The documented design of that `preserveAspectRatio` should use the latest near but not larger value of one edge. So current these code:

```objctivec
            if (pixelRatio > thumbnailRatio) {
                maxPixelSize = thumbnailSize.width;
            } else {
                maxPixelSize = thumbnailSize.height;
            }
```

will only assume to use another edge's length as input thumbnail pixel width. Which will make the final thumbnail smaller than you expected.

**Comment by dreampiggy** [comment]
Fixed in [5.12.6](https://github.com/rs/SDWebImage/releases/tag/5.12.6)
